### PR TITLE
Make help documentation accessible offline #417

### LINKED ISSUE #417: Make help documentation accessible offline
The help window displays an online version of the documentation at https://se-edu.github.io/addressbook-level4/docs/UserGuide.

However, the documentation is always updated, so it might not match the version that the user is currently using. Also, the user may not always have an internet connection to access the documentation.

Therefore, we should make it offline.

The help documentation should ideally be saved inside `src/main/resources/` folder so that:
* It will automatically be bundled when doing `shadowJar`.
* It will be easier to reference the URL of the documentation.

(Reference: https://github.com/CS2103AUG2016-W11-C4/main/commit/5d3dfcb76adffc33710c95820644b6d7d5a081a6)

### LINKED ISSUE #419: Add custom gradle task for generating offline UserGuide files
Prerequisite: #417 

Manually updating the offline UserGuide files would be tedious and error prone. We should write a custom gradle task that automatically generates the offline UserGuide when releasing a new version.

### PR DESCRIPTION
**Pull request by yamgent**
Fixes #417.
Also fixes #419.

Proposed merge commit message:

```
The help window accesses the online version of the documentation.

Users are unable to access it if they are offline. Furthermore, the
application version for the online documentation may not match the
application version that the user is using.

Let's change the help window to use an offline version instead, and
deploy the offline version to the 'src/main/resources/' folder such 
that shadowJar will automatically bundle the documentation with the app.
```

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@yamgent submitted v1 for review.

* [v1 1/1] [HelpWindow: Use offline version of user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/d4f0da27cdc572f7bfe3166d14303b3a9fc82ee4)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/ef06553753b5073c012ccac9ab1082c46b9d0f51...d4f0da27cdc572f7bfe3166d14303b3a9fc82ee4))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/420/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by yamgent** [comment]
@Zhiyuan-Amos can help to check if the HelpWindow still display gibberish text on Mac for this PR? (with reference to #374)

**Comment by Zhiyuan-Amos** [comment]
Your PR fixes it.

File: src/main/java/seedu/address/ui/HelpWindow.java [hunk 1]
```diff
@@ -35,7 +35,8 @@ public HelpWindow() {
         dialogStage.setMaximized(true); //TODO: set a more appropriate initial size
         FxViewUtil.setStageIcon(dialogStage, ICON);
 
-        browser.getEngine().load(USERGUIDE_URL);
+        URL userGuideUrl = getClass().getResource(USERGUIDE_FILE_PATH);
+        browser.getEngine().load(userGuideUrl.toString());
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/HelpWindow.java]: bring `toString()` up, since `URL userGuideUrl` will not be used in other places except here.

**Comment by CanIHasReview** [comment]
# v2

@yamgent submitted v2 for review.

* [v2 1/1] [HelpWindow: Use offline version of user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/964a95c3482b09dff93191db08f6a9fb8062f810)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/ef06553753b5073c012ccac9ab1082c46b9d0f51...964a95c3482b09dff93191db08f6a9fb8062f810)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/6ca9bb1a83efb2930e552f765e0e79a117d81238~1...6ca9bb1a83efb2930e552f765e0e79a117d81238))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/420/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/resources/docs/UserGuide.html [hunk 2]
```diff
@@ -0,0 +1,268 @@
+<!DOCTYPE html>
```
Comment (damithc) [inline src/main/resources/docs/UserGuide.html]: Is this file auto-generated by gradle?

Comment (yamgent) [inline src/main/resources/docs/UserGuide.html]: No, there's no corresponding asciidoc document for UserGuide.md yet, so we cannot generate a html file yet. The other PR by @limmlingg only has the Developer Guide atm.

**Comment by CanIHasReview** [comment]
# v3

@yamgent submitted v3 for review.

* [v3 1/1] [HelpWindow: Use offline version of user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/6997c5dc7d07d04570ee33e9637356cb56982771)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cc22ed642b34f15798facec1bddd597aa8c0c395...6997c5dc7d07d04570ee33e9637356cb56982771)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/93e99b27581749901232fc890a7b1e98a511bc0a~1...93e99b27581749901232fc890a7b1e98a511bc0a))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/420/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
@yamgent are we putting this on hold until the asciidoc version of the user guide is available?

**Comment by yamgent** [comment]
We can put this on hold, not that urgent currently.

**Comment by yamgent** [comment]
@eugenepeh I need your help. For some reasons, I am unable to get checkstyle to ignore the 'src/main/resources/docs' folder, so Travis keep failing for this PR (I can't reproduce the errors on my local laptop).

**Comment by eugenepeh** [comment]
@yamgent got it, will check it in a bit

**Comment by eugenepeh** [comment]
@yamgent the error doesnt seems to come from checkstyle as I removed the checkstyleMain and checkstyleTest from travis gradle build and it still fails 
https://travis-ci.org/CS2103R-Eugene-Peh/addressbook-level4/builds/237144841?utm_source=github_status&utm_medium=notification

**Comment by yamgent** [comment]
@eugenepeh Thanks for the help, appreciate it! :) Will take a look at how this can be resolved.

**Comment by pyokagan** [comment]
A quick look at the diff shows that `src/main/resources/docs/UserGuide.html`and `src/main/resources/docs/stylesheets/coderay-asciidoctor.css` do indeed not have newline at EOF.

Is there a way to *not* check build products into the repo?

**Comment by yamgent** [comment]
My thinking was that we would have to modify this script (https://github.com/se-edu/addressbook-level4/blob/master/config/travis/check-eof-newline.sh) to teach it not to check for build products inside the source code directory.

**Comment by pyokagan** [comment]
For clarity: I'm saying that build products should *not* be checked into the repo. Instead, they should be generated as part of the build process (hook onto the `processResources` gradle task or something)

**Comment by yamgent** [comment]
> For clarity: I'm saying that build products should not be checked into the repo. Instead, they should be generated as part of the build process (hook onto the processResources gradle task or something)

Got it, thanks. It currently only generates when doing a `shadowJar`. I will look at `processResources`, and also update `.gitignore` to prevent these offline documentation files from being checked in.

**Comment by CanIHasReview** [comment]
# v4

@yamgent submitted v4 for review.

* [v4 1/2] [Create gradle task for generating offline help documentation](https://github.com/se-edu/addressbook-level4/pull/420/commits/2fc3218b71cb96cfd907dfea5efb7c32704875cc)
* [v4 2/2] [HelpWindow: Use offline version of user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/a49d0659b8711bbb22b345273938fb4c89d0c690)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...a49d0659b8711bbb22b345273938fb4c89d0c690)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/47772f127cbd9edac6d4f75bba85966b2779c49d~1...47772f127cbd9edac6d4f75bba85966b2779c49d))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/420/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: build.gradle [hunk 3]
```diff
@@ -4,7 +4,6 @@
  * For more details take a look at the Java Quickstart chapter in the Gradle
  * user guide available at http://gradle.org/docs/2.2.1/userguide/tutorial_java_projects.html
  */
-
```
Comment (Zhiyuan-Amos) [inline build.gradle]: random deletion of line

File: docs/UsingGradle.md [hunk 4]
```diff
@@ -52,6 +52,9 @@ If we package only our own class files into the JAR file, it will not work prope
 * **`asciidoctor`**<br>
   Converts AsciiDoc files in `docs` to HTML format. Generated HTML files can be found in `build/docs`.
 
+* **`deployOfflineDocs`**<br>
+  Updates the offline documentation used by the Help window in the application.
```
Comment (limmlingg) [inline docs/UsingGradle.md]: Should we be more specific and say `User Guide` instead of `documentation`?

**Comment by CanIHasReview** [comment]
# v5

@yamgent submitted v5 for review.

* [v5 1/2] [Create gradle task for generating offline user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/3b734b17ceb90785b3b60cc7d0ac7501f72926a3)
* [v5 2/2] [HelpWindow: Use offline version of user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/03b2b3825a0c21e48746c59576188684fa5a1e05)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...03b2b3825a0c21e48746c59576188684fa5a1e05)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/fc4717c479d4b0b87197434815971b9e94871b42~1...fc4717c479d4b0b87197434815971b9e94871b42))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/420/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v6

@yamgent submitted v6 for review.

* [v6 1/2] [Create gradle task for generating offline user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/57de466738199b49e0fb5ac23f6cd5c6dc0eaaa6)
* [v6 2/2] [HelpWindow: Use offline version of user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/b7b73498ce319fb17207ac114492cac16f462489)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/d8b24e8a0c61af4746f7dce842eb2771c8583402...b7b73498ce319fb17207ac114492cac16f462489)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/ec25837a687c7d1a2378872fa692de32e9757170~1...ec25837a687c7d1a2378872fa692de32e9757170))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/420/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by yamgent** [comment]
Stylesheets seem to not load when exported to .jar. Will explore this further.

help

**Comment by yamgent** [comment]
Exporting to jar:

* Windows: Everything works perfectly fine.
* macOS: Everything works perfectly fine (although gibberish text, as reported in #374, reappears).
* Linux: The page loads but no images and stylesheets.

I suspect that JavaFx, for some reason, fail to resolve the relative path for the resource properly in Linux. I tried replacing the path with an absolute one. For example:

`Ui`

This works, but doing absolute path is obviously a very bad idea.

Previously there were some css and resource files bugs for JavaFx's WebView on all platforms, perhaps it got fixed on all platforms except Linux? Can't find any additional info on the web though.

@damithc Not sure how to proceed from here.

**Comment by damithc** [comment]
As discussed, options:
1. Confirm it's a known bug and there are no workarounds. In that case we may have to go back to online docs.
2. We may be doing something wrong. In that case need to debug and fix.

If problematic for both Mac and Linux, we may have to go back to online docs and use something like rawgit to serve different versions of the docs to different versions of the app.

**Comment by yamgent** [comment]
I created an new project that completely isolates the problem (in case anyone's interested: [Source Code](https://drive.google.com/open?id=0BzKsUiaw_rSSekp5ejNNQnVjZm8), [.jar File](https://drive.google.com/open?id=0BzKsUiaw_rSSZldqSF91UnZkMlU)).

I still face the same issue even with the new project (it does not show the text "Hello World" blue and bolded when deployed to jar, which clearly indicates that the .css file is not being properly loaded).

Therefore, I filed a bug report [here](http://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8181642). The verdict: it looks like they are unable to reproduce the bug at all using the exact same source code and .jar file on their Linux box.

At this stage, I conclude that it is just the problem on my side. I will keep looking, but if I am the only one having this problem, it doesn't seem necessary to delay this PR just because of that.

On a side note:

> If problematic for both Mac and Linux, we may have to go back to online docs and use something like rawgit

There was never a problem with the css file on the Mac side (with regards to the gibberish text, both online and offline exhibit this problem so it does not matter either way).

**Comment by pyokagan** [comment]
Styling seems to work on my box (linux).

stylingworks

**Comment by pyokagan** [comment]
> Styling seems to work on my box (linux).

I ran `./gradlew runShadow` on this PR branch.

**Comment by yamgent** [comment]
Thanks for confirming @pyokagan, will triple confirm with teammates people, then I will rebase and it is ready for merge.

**Comment by CanIHasReview** [comment]
# v7

@yamgent submitted v7 for review.

* [v7 1/2] [Create gradle task for generating offline user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/591c4c521c592d0008f67be98adf9e26ee1fb2f9)
* [v7 2/2] [HelpWindow: Use offline version of user guide](https://github.com/se-edu/addressbook-level4/pull/420/commits/ae3f81e3ea549be88aac672f9498ee599015c45a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a76a2a505499c1c7c68a22083026f286f4f85592...ae3f81e3ea549be88aac672f9498ee599015c45a)) ([:chart_with_upwards_trend: Interdiff between v6 and v7](https://github.com/se-edu/addressbook-level4/compare/8e8a4c61c27b850210175723a8471c205b6455b4~1...8e8a4c61c27b850210175723a8471c205b6455b4))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/420/7/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Pull request merged**